src/main/scala/com/cognite/spark/datasource/EventsRelation.scala

```diff
@@ -136,13 +139,111 @@ class EventsRelation(config: RelationConfig)(@transient val sqlContext: SQLConte
+          s"Filtering using 'string ends with' not allowed for events, attempted for ${value.toString}")
+      case StringContains("type", value) =>
+        sys.error(
+          s"Filtering using 'string contains' not allowed for data points, attempted for ${value.toString}")
+      case _ =>
+        Seq()
+    }
+
+  sealed case class SubTypeFilter(subType: String)
+
+  def getSubTypeFilters(filter: Filter): Seq[SubTypeFilter] =
+    filter match {
+      case IsNotNull("subtype") => Seq()
+      case EqualTo("subtype", value) => Seq(SubTypeFilter(value.toString))
+      case EqualNullSafe("subtype", value) => Seq(SubTypeFilter(value.toString))
+      case In("subtype", values) => values.map(v => SubTypeFilter(v.toString))
+      case And(f1, f2) => getSubTypeFilters(f1) ++ getSubTypeFilters(f2)
+      case Or(f1, f2) => getSubTypeFilters(f1) ++ getSubTypeFilters(f2)
+      // TODO: add support for String filtering using the 'Search for events' endpoint
+      case StringStartsWith("subtype", value) =>
+        sys.error(
+          s"Filtering using 'string starts with' not allowed for events, attempted for ${value.toString}")
+      case StringEndsWith("subtype", value) =>
+        sys.error(
+          s"Filtering using 'string ends with' not allowed for events, attempted for ${value.toString}")
+      case StringContains("subtype", value) =>
+        sys.error(
+          s"Filtering using 'string contains' not allowed for data points, attempted for ${value.toString}")
+      case _ =>
+        Seq()
+    }
+  // scalastyle:on cyclomatic.complexity
+  override def buildScan(): RDD[Row] = buildScan(Array.empty, Array.empty)
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val types = filters.flatMap(getTypeFilters).map(_.`type`)
+    val subTypes = filters.flatMap(getSubTypeFilters).map(_.subType)
+
+    val params = Map("type" -> types.toString, "subtype" -> subTypes.toString)
+
+    val rdds: Array[CdpRdd[EventItem]] = for {
+      eventType <- if (types.isEmpty) { Array(None) } else { types }
+      eventSubType <- if (subTypes.isEmpty) { Array(None) } else { subTypes }
+    } yield {
+      val urlWithType =
+        eventType match {
+          case t: String => listUrl(config).param("type", t.toString)
+          case None => listUrl(config)
+        }
+
+      val urlWithTypeAndSubType = {
+        eventSubType match {
+          case et: String => urlWithType.param("subtype", et.toString)
+          case None => urlWithType
+        }
+      }
+
+      CdpRdd[EventItem](
+        sqlContext.sparkContext,
+        (e: EventItem) => {
+          if (config.collectMetrics) {
+            eventsRead.inc()
+          }
+          toRow(e, requiredColumns)
+        },
+        urlWithTypeAndSubType.param("onlyCursors", "true"),
+        urlWithTypeAndSubType,
+        config
+      )
+    }
+
+    rdds.foldLeft(sqlContext.sparkContext.emptyRDD[Row])((a, b) => a.union(b))
```

Comment:
We should try to get rid of those `.unions`, they create potentially large DAGs that I suspect may have a performance impact. Can probably be fixed by having smarter custom `RDD` implementations - we shoudln't fix it here and now but I just wanted to point this out.